### PR TITLE
Second preimage attack prevention

### PR DESCRIPTION
**Pull request by lontivero**
This is the simplest way I found to make the length of the messages part of the message itself. This is not exactly how TupleHash is defined here: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf but it is easy to do anyway, we just need to implement `encode_left` and `encode_right` and also concatenate the length of the final string at the end of the string. I can implement it if we find it worthy.

### CODE REVIEW

**Review by nothingmuch — APPROVED** [review]
This addresses any concerns of variable length fields potentially being added at some later point, but does not yet add any commitment to the statement (I think HashToScalar would just need to to also commit to a unique tag for the statement), or in the case of conjunctions, to transcripts of prior sub-statements.

I'm still in favor of using STROBE because it gives conjunction/AND for free and being a standard would be easier to audit when considering potentially multiple implementations in different languages. i can't think of anything that would really go wrong with this commit's approach specifically, but STROBE provides a clean and principled way of doing these things (see the whole talk i linked to for a good overview).

Conjunctions are free because the prover and verifier just plug statements, public inputs, and any other required metadata such as static parameters, type tags, etc into the STROBE transcript object in the same order, and whenever a challenge is needed STROBE can just output a hash of the current state comitting to all previous states at any point, binding the challenges of subsequent proofs to those appearing earlier in the transcript.

This would be valuable for implementing the range proofs, especially if we decide we to go with bulletproofs for the size reduction (which I suspect will be significant if long term fee credentials are used) by automatically ensuring that proofs validate only if the entire transcript is identical, without needing to define or implement ways to add hashes of previous states to the hashing of related proofs in some ad-hoc way.

**Comment by lontivero** [comment]
I didn't read the STROBE paper, just scrolled the pdf and if you think we can implement the STROBE **Protocol transcripts** 
 then I agree. On the other hand, if your idea is to use/implement more than that then I think that's not so easy for us. Wasabi uses the same cryptography used by bitcoin, I mean, same hash algorithms, same curve and basically same everything. This allow us to use NBitcoin and NBitcoin.Secp256k1 libraries without having to add other dependencies (dotnet or not). STROBE seems to use Curve25519 curve for example what I am sure we don't want to have.

**Comment by nothingmuch** [comment]
i think you're thinking of Noise handshakes, STROBE requires just a single
primitive, extensible output hash functions.

using only sha-256 something similar could be accomplished using Merkle
roots of transcripts as challenges, STROBE is a simpler way of achieving
the same effect using just SHAKE, which is one of the SHA-3 hash functions,
so this should be simpler than designing and implementing something new,
both because it's a simpler construction that had been studied and intended
for this purpose.

SHAKE itself could be copied from bouncy castle which includes it, and
there is also an educational STROBE csharp port we can study, we only need
to implement a few of the STROBE operations instead of coming up with a way
of doing this on our own.

basically any two elementary proofs appearing as subproofs in a conjunction
need to use challenges that include each others' commitments in their
transcript hashes, not just their own, in order to produce as output a
proof of the conjugation and only that conjunction. it's intuitive to
implement with sha-256 but i think more complex and risky overall.

**Comment by nothingmuch** [comment]
Another good reason: it protects against bad RNGs: https://merlin.cool/transcript/rng.html

Suppose a client has a broken RNG and proves the same credential twice to the server with reused nonces. in that case I think the z witness would be revealed completely deanonymizing them, and the server woulnd't even need to work hard to collide all credential presentations since the serial numbers would be the same.

**Comment by nopara73** [comment]
@lontivero @nothingmuch please discuss and make another PR if you want to do something else here.
